### PR TITLE
sorter: Unified Sorter clean up stale temporary files

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve? 
- Unified Sorter is not able to clean up its temporary files if the process exits abnormally. If the process OOMs frequently, disk space can be exhausted and will require a manual clean-up.

### What is changed and how it works?
- File lock-based stale file detection and clean up.
- Prevents multiple `cdc server` instances from sharing the same `sort-dir`.

### Check List 

Tests 

 - Unit test
 - Integration test 

Code changes

 - Has persistent data change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to modify TiDB operator to make TiCDC a stateful set.

### Release note

- Add stale temporary files clean-up in Unified Sorter, and forbids sharing sort-dir.

### CODE REVIEW

File: pkg/filelock/filelock.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+// FileLock represents a file lock created by `flock`.
+// For more on the system call `flock`, read:
+// https://linux.die.net/man/2/flock
+type FileLock struct {
+	fd *os.File
+}
+
+// NewFileLock creates a new file lock on the file described in filePath.
+func NewFileLock(filePath string) (*FileLock, error) {
+	file, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE|syscall.O_NONBLOCK, 0o666)
```
Comment (overvenus) [inline pkg/filelock/filelock.go:33]: Please keep minimal permission requirements.
```suggestion
	file, err := os.OpenFile(filePath, os.O_RDWR|os.O_CREATE|syscall.O_NONBLOCK, 0o600)
```

File: pkg/filelock/simple_filelock.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+// SimpleFileLock represents a file used as a lock.
+type SimpleFileLock struct {
+	filePath string
+}
+
+// NewSimpleFileLock creates a new SimpleFileLock by filePath
+func NewSimpleFileLock(filePath string) (*SimpleFileLock, error) {
+	startTime := time.Now()
+	err := retry.Run(time.Millisecond*50, 20, func() error {
```
Comment (overvenus) [inline pkg/filelock/simple_filelock.go:39]: Why does it need retry?

File: pkg/filelock/simple_filelock.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+		lockFile, err := os.OpenFile(filePath, os.O_CREATE|os.O_EXCL|os.O_RDWR, 0o666)
```
Comment (overvenus) [inline pkg/filelock/simple_filelock.go:40]:
```suggestion
		lockFile, err := os.OpenFile(filePath, os.O_CREATE|os.O_EXCL|os.O_RDWR, 0o600)
```

File: pkg/filelock/simple_filelock.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &SimpleFileLock{filePath}, nil
+}
+
+// Unlock unlocks the SimpleFileLock
+func (fl *SimpleFileLock) Unlock() error {
+	err := os.Remove(fl.filePath)
```
Comment (overvenus) [inline pkg/filelock/simple_filelock.go:69]: Why does it need to remove the file?

File: pkg/filelock/simple_filelock.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/pingcap/ticdc/pkg/retry"
+	"go.uber.org/zap"
+)
+
+const (
+	cleanLockInterval = 10 * time.Second
+)
+
+// SimpleFileLock represents a file used as a lock.
+type SimpleFileLock struct {
```
Comment (overvenus) [inline pkg/filelock/simple_filelock.go:32]: What's the difference between `SimpleFileLock` and `FileLock`?

File: cdc/puller/sorter/backend_pool.go [hunk 6]
```diff
@@ -283,3 +308,127 @@ func (p *backEndPool) memoryPressure() int32 {
...
+		return errors.Trace(err)
+	}
+
+	log.Info("sorter temporary file prefix unlocked", zap.String("prefix", p.filePrefix))
+	return nil
+}
+
+func (p *backEndPool) lockPrefix() error {
+	metaLockPath := fmt.Sprintf("%s/cdc-meta-lock", p.dir)
+	localLockPath := fmt.Sprintf("%slock", p.filePrefix)
```
Comment (overvenus) [inline cdc/puller/sorter/backend_pool.go:341]: Why does it need to hold two locks?

File: cdc/puller/sorter/file_backend.go [hunk 7]
```diff
@@ -165,13 +214,20 @@ func (r *fileBackEndReader) readNext() (*model.PolymorphicEvent, error) {
 	if err != nil {
 		if err == io.EOF {
 			r.isEOF = true
+			// verifies that the file has not been truncated unexpectedly.
```
Comment (amyangfei) [inline cdc/puller/sorter/file_backend.go:217]: Does the truncate here only happens by accident, not by cdc itself

File: cdc/puller/sorter/backend_pool.go [hunk 8]
```diff
@@ -71,6 +74,17 @@ func newBackEndPool(dir string, captureAddr string) *backEndPool {
 		filePrefix:        fmt.Sprintf("%s/sort-%d-", dir, os.Getpid()),
 	}
 
+	err := ret.lockSortDir()
+	if err != nil {
+		log.Error("failed to lock file prefix", zap.String("prefix", ret.filePrefix))
+		return nil
+	}
```
Comment (amyangfei) [inline cdc/puller/sorter/backend_pool.go:81]: The unified sorter would not work if lock prefix failed and backend pool is nil. Besides the error is not logged anywhere

File: cdc/puller/sorter/backend_pool.go [hunk 9]
```diff
@@ -283,3 +307,63 @@ func (p *backEndPool) memoryPressure() int32 {
...
+	return nil
+}
+
+func (p *backEndPool) cleanUpStaleFiles() error {
+	if p.dir == "" {
+		// guard against programmer error. Must be careful since we are deleting user files.
+		log.Panic("unexpected file prefix", zap.String("file-prefix", p.filePrefix))
+	}
+
+	files, err := filepath.Glob(p.dir + "/sort-*")
```
Comment (amyangfei) [inline cdc/puller/sorter/backend_pool.go:349]:
```suggestion
	files, err := filepath.Glob(filepath.Join(p.dir, "sort-*"))
```

File: cdc/puller/sorter/backend_pool.go (hunk 8 above)
Comment (liuzix) [reply cdc/puller/sorter/backend_pool.go:81]: will fix later

File: cdc/puller/sorter/file_backend.go (hunk 7 above)
Comment (liuzix) [reply cdc/puller/sorter/file_backend.go:217]: CDC will truncate a file once its content is no longer needed and the file is going to be recycled. Normally a file in use is not truncated. This piece of code here verifies against accidental truncation either by CDC bug or by user misoperations.

File: pkg/filelock/simple_filelock.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+	filePath string
+}
+
+// NewSimpleFileLock creates a new SimpleFileLock by filePath
+func NewSimpleFileLock(filePath string) (*SimpleFileLock, error) {
```
Comment (overvenus) [inline pkg/filelock/simple_filelock.go:37]: It seems dead code in the current implementation.

**Review by overvenus — COMMENTED** [review]
/lgtm

**Pull request merged**